src/rust/engine/options/src/parse.rs

```diff
@@ -274,49 +278,188 @@ fn format_parse_error(
+// or None if the file doesn't exist and the @?fromfile syntax was used.
+//
+// Otherwise, the first component is None and the second is the original value.
+type ExpandedValue = (Option<PathBuf>, Option<String>);
+
+fn mk_parse_err(err: impl Display, path: &Path) -> ParseError {
+    ParseError::new(format!(
+        "Problem reading {path} for {{name}}: {err_msg}",
+        path = path.display(),
+        err_msg = err,
+    ))
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_bool_list(value: &str) -> Result<Vec<ListEdit<bool>>, ParseError> {
-    option_value_parser::bool_list_edits(value)
-        .map_err(|e| format_parse_error("bool list", value, e))
+fn maybe_expand(value: String) -> Result<ExpandedValue, ParseError> {
+    if let Some(suffix) = value.strip_prefix('@') {
+        if suffix.starts_with('@') {
+            // @@ escapes the initial @.
+            Ok((None, Some(suffix.to_owned())))
+        } else {
+            match suffix.strip_prefix('?') {
+                Some(subsuffix) => {
+                    // @? means the path is allowed to not exist.
+                    let path = PathBuf::from(subsuffix);
+                    match fs::read_to_string(&path) {
+                        Ok(content) => Ok((Some(path), Some(content))),
+                        Err(err) if err.kind() == io::ErrorKind::NotFound => Ok((Some(path), None)),
+                        Err(err) => Err(mk_parse_err(err, &path)),
+                    }
+                }
+                _ => {
+                    let path = PathBuf::from(suffix);
+                    let content = fs::read_to_string(&path).map_err(|e| mk_parse_err(e, &path))?;
+                    Ok((Some(path), Some(content)))
+                }
+            }
+        }
+    } else {
+        Ok((None, Some(value)))
+    }
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_int_list(value: &str) -> Result<Vec<ListEdit<i64>>, ParseError> {
-    option_value_parser::int_list_edits(value).map_err(|e| format_parse_error("int list", value, e))
+pub(crate) fn expand(value: String) -> Result<Option<String>, ParseError> {
+    let (_, expanded_value) = maybe_expand(value)?;
+    Ok(expanded_value)
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_float_list(value: &str) -> Result<Vec<ListEdit<f64>>, ParseError> {
-    option_value_parser::float_list_edits(value)
-        .map_err(|e| format_parse_error("float list", value, e))
+pub(crate) fn expand_to_list<T: ListMember>(
+    value: String,
+) -> Result<Option<Vec<ListEdit<T>>>, ParseError> {
+    let (path_opt, value_opt) = maybe_expand(value)?;
+    if value_opt.is_none() {
+        return Ok(None);
+    }
+    let value = value_opt.unwrap();
+    let mut deserialized_items: Option<Vec<T>> = None;
+    if let Some(path) = path_opt {
+        if let Some(ext) = path.extension() {
+            if ext == "json" {
+                deserialized_items =
+                    Some(serde_json::from_str(&value).map_err(|e| mk_parse_err(e, &path))?)
```

Comment:
serde is freakin' magic! This (plus a minor annotation on the Val class in lib.rs) is all I needed to get Rust to parse json and yaml into our existing data structures!